README.md
# UniPagerLED
Front panel status LED

A software that connects to the websocket interface of RustPager and controls front panel LED with function
* UniPager running (means Websocket connection is up and running)
* Connected to DAPNET-Core
* Transmitting

Configuration parameters on the command line are:
* Websocket Hostname (Default: localhost) --hostname
* Websocket Port (Default: 8055) --port
* GPIO Pin for "Running LED"  --gpioRun
* GPIO Pin for "Connected LED" --gpioConn
* GPIO Pin for "Transmitting LED" --gpioTX
  * The GPIOs can be inverted to be flexible if the LEDs common Pin is Ground or 3.3V by putting a __-__ in front of the pin number; e.g. --gpioTX -29
* Presets for the LEDs for well known hardware --preset
  * --preset help gives a list of all known presets

The pin header numbers have top be used, not WiringPi Numbers.

If no pin config is given it tries to guess the config based on the transmittertype supplied by the transmitter config. Currently, only the C9000 with the preset below is supported.

# Common assigments:
__RasPager9000:__
* gpioTX: none
* gpioRun: 24
* gpioConn: 26
* preset: c9000


Websocket connection should be opened once and kept open. React on update comming from the UniPager. If connection is closed, reconnect.

Packages needed for the Python Program:
* sudo apt install python3 python3-websocket python3-rpi.gpio


debian/changelog
unipager-led (1.0.0) unstable; urgency=medium

  * Initial Release

 -- RWTH Amateurfunkgruppe <rwth-afu@online.de>  Thu, 30 Mar 2017 12:00:00 +0200


src/main.cpp
#include "client_ws.hpp"
#include <wiringPi.h>


using namespace std;

typedef SimpleWeb::SocketClient<SimpleWeb::WS> WsClient;

int main() {
    
    //Example 4: Client communication with server
    //Possible output:
    //Server: Opened connection 140184920260656
    //Client: Opened connection
    //Client: Sending message: "Hello"
    //Server: Message received: "Hello" from 140184920260656
    //Server: Sending message "Hello" to 140184920260656
    //Client: Message received: "Hello"
    //Client: Sending close connection
    //Server: Closed connection 140184920260656 with status code 1000
    //Client: Closed connection with status code 1000
    WsClient client("localhost:8055");
    client.on_message=[&client](shared_ptr<WsClient::Message> message) {
        auto message_str=message->string();
        
        cout << "Client: Message received: \"" << message_str << "\"" << endl;
        
        cout << "Client: Sending close connection" << endl;
        client.send_close(1000);
    };
    
    client.on_open=[&client]() {
        cout << "Client: Opened connection" << endl;
        
        string message="GetVersion";
        cout << "Client: Sending message: \"" << message << "\"" << endl;

        auto send_stream=make_shared<WsClient::SendStream>();
        *send_stream << message;
        client.send(send_stream);
    };
    
    client.on_close=[](int status, const string& /*reason*/) {
        cout << "Client: Closed connection with status code " << status << endl;
    };
    
    //See http://www.boost.org/doc/libs/1_55_0/doc/html/boost_asio/reference.html, Error Codes for error code meanings
    client.on_error=[](const boost::system::error_code& ec) {
        cout << "Client: Error: " << ec << ", error message: " << ec.message() << endl;
    };
    
    client.start();

    return 0;
}


statusleds.py
#!/usr/bin/env python3

import websocket
import json
import time
import argparse
import signal
import sys
import RPi.GPIO as GPIO

presets = {"c9000": {
	"run": 24,
	"conn": 26,
	"tx": None
}}

presetconfigmap = {"C9000":"c9000"}

class Statusleds:
	def __init__(self, url, runled, connled, txled):
		self.runled = runled
		self.connled = connled
		self.txled = txled
		
		self.ws = websocket.create_connection(url, timeout=10)
		
		if runled == connled == txled == None:
			self.ws.send('"GetConfig"')
			sconfig = self.ws.recv()
			config = json.loads(sconfig)
			print("Config:")
			print(json.dumps(config,indent=4))
			transmitter = config["Config"]["transmitter"]
			if transmitter in presetconfigmap:
				self.runled = presets[presetconfigmap[transmitter]]["run"]
				self.connled = presets[presetconfigmap[transmitter]]["conn"]
				self.txled = presets[presetconfigmap[transmitter]]["tx"]
		
		GPIO.setmode(GPIO.BOARD)
		if not self.runled is None:
			GPIO.setup(abs(self.runled), GPIO.OUT)
		if not self.connled is None:
			GPIO.setup(abs(self.connled), GPIO.OUT)
		if not self.txled is None:
			GPIO.setup(abs(self.txled), GPIO.OUT)
		
		self.setled(self.runled, True)
	
	def __enter__(self):
		return self
	
	def __exit__(self, exc_type, exc_value, traceback):
		self.setled(self.runled, False)
		GPIO.cleanup()
		self.ws.close()
	
	def setled(self, led, status):
		if led == None:
			print("Led not present")
			return
		if led < 0:
			led = -led
			status = not status
		print("Setting led %d to %d" % (led, status))
		GPIO.output(led, status)
	
	def getstatus(self):
		self.ws.send('"GetStatus"')

	def setstatus(self):
		sres = self.ws.recv()
		res = json.loads(sres)
		print("Resp:")
		print(json.dumps(res,indent=4))
		try:
			status = res["Status"]
		except KeyError:
			print("Other message, ignoring")
			return
		self.setled(self.connled, status["connected"])
		self.setled(self.txled, status["transmitting"])
	
	def loop(self):
		self.getstatus()
		while True:
			self.setstatus()

parser = argparse.ArgumentParser(description='Set status LEDs')
parser.add_argument('--hostname', default='localhost',
                    help='The host running RustPager, default localhost')
parser.add_argument('--port', default='8055',
                    help='The port RustPager is listening, default 8055')
parser.add_argument('--gpioRun', dest='runled', default=None, type=int,
                    help='UniPager running')
parser.add_argument('--gpioConn', dest='connled', default=None, type=int,
                    help='DAPNET connection ok')
parser.add_argument('--gpioTX', dest='txled', default=None, type=int,
                    help='Transmitting')
parser.add_argument('--preset', dest='preset', default=None, type=str,
                    help='Preset, help for help')

args = parser.parse_args()

preset = args.preset

if not preset is None:
	if preset == "help":
		print("Presets:")
		for pr in presets:
			print("%s:" %pr)
			for t in ["run","conn","tx"]:
				print("  %s: %s" %(t, presets[pr][t]))
		sys.exit(0)
	elif preset in presets:
		runled = presets[preset]["run"]
		connled = presets[preset]["conn"]
		txled = presets[preset]["tx"]
	else:
		print("Preset %s unknown" %preset)
		sys.exit(1)
else:
	runled = args.runled
	connled = args.connled
	txled = args.txled

while True:
	try:
		with Statusleds("ws://%s:%s/" %(args.hostname, args.port), runled, connled, txled) as setter:

			setter.loop()
	except websocket._exceptions.WebSocketTimeoutException:
		print("Timeout")
	except websocket._exceptions.WebSocketConnectionClosedException:
		print("Websocket closed")
	except ConnectionRefusedError:
		print("Connection Refused")
	except KeyboardInterrupt:
		print("Shutting down")
		sys.exit(0)
	
	print("Waiting 10 seconds")
	try:
		time.sleep(10)
	except KeyboardInterrupt:
		print("Shutting down")
		sys.exit(0)
	print("Trying again")

